internal/pkg/manifest/env.go

```diff
@@ -102,9 +109,107 @@ type environmentNetworkConfig struct {
 }
 
 type environmentVPCConfig struct {
-	ID      *string              `yaml:"id,omitempty"`
-	CIDR    *IPNet               `yaml:"cidr,omitempty"`
-	Subnets subnetsConfiguration `yaml:"subnets,omitempty"`
+	ID                  *string              `yaml:"id,omitempty"`
+	CIDR                *IPNet               `yaml:"cidr,omitempty"`
+	Subnets             subnetsConfiguration `yaml:"subnets,omitempty"`
+	SecurityGroupConfig securityGroupConfig  `yaml:"security_group,omitempty"`
+}
+
+type securityGroupConfig struct {
+	Ingress []SecurityGroupRule `yaml:"ingress,omitempty"`
+	Egress  []SecurityGroupRule `yaml:"egress,omitempty"`
+}
+
+func (cfg securityGroupConfig) isEmpty() bool {
+	return len(cfg.Ingress) == 0 && len(cfg.Egress) == 0
+}
+
+// SecurityGroupRule holds the security group ingress and egress configs.
+type SecurityGroupRule struct {
+	CidrIP      string      `yaml:"cidr"`
+	PortsConfig PortsConfig `yaml:"ports"`
+	IpProtocol  string      `yaml:"ip_protocol"`
+}
+
+// PortsConfig represents a range of ports [from:to] inclusive.
+// The simple form allow represent from and to ports as a single value, whereas the advanced form is for different values.
+type PortsConfig struct {
+	Port  *int            // 0 is a valid value, so we want the default value to be nil.
+	Ports *PortsRangeBand // Mutually exclusive with port.
+}
+
+// PortsRangeBand represents a range of from and to port values.
+// For example, "0-65536".
+type PortsRangeBand string
```

Comment:
Very good question, even I was wondering if I need these structs to be exported but then I could see almost all the structs are kept public in env manifest file. But you are right I should change them if they are not exported. 

Also I did not use `IntRangeBand` purposefully because `IntRangeBand` has different definition of `validate` than what we needed for the port configs.

Reply:
I'm not sure which part of `IntRangeBand`'s [`validate` logic](https://github.com/aws/copilot-cli/blob/f04f52f4e856bbb12687720594b9ced2d15ba329/internal/pkg/manifest/validate.go#L1029) can't be reused here 🤔 I might be missing something obvious, but it seems to me that both struct's `validate` are just checking if they are formatted as `min-max`...

Reply:
So the [error message](https://github.com/aws/copilot-cli/blob/f04f52f4e856bbb12687720594b9ced2d15ba329/internal/pkg/manifest/validate.go#L1034) that validate function returns is different than what we need. Customer might get confused by the min and max wordings in the error message. Other than that I guess everything is fine. It's just that we return human readable errors for `strconv.Atoi`.  Let me know if you still want me to use `IntRangeBand`